Fix pump timeout

benguillet opened the pull request:
The old architecture was blocking in some conditions and thus triggering a pump timeout in logspout, which results in logspout deleting the logstream channel and completely stoppping to send logs to the adapter.

This is a refactor using channels, select and better goroutines, so that it never blocks. If the flush takes too long and the flusher start to accumulate delay (more than 10 buffers in the queue), it starts dropping buffers (with a warning). This refactor also adds tests.

benguillet committed:
Remove useless check for open channel and FlushAll.

Logspout will delete the channel from its internal map before
closing it anyway, in case of timeout. So this code will never run.

benguillet committed:
Perfom buffer.Add() in a goroutine.

This allow the adapter.Stream function to be extremely fast and prevents
timeouts from the pump. Before, the pump would just delete the logstream
channel from its mapping, and logs would stop being send to the adapter in
anything was stuck in Stream for more than 1 second.

ejholmes commented on kinesis.go:
```diff
@@ -60,7 +55,7 @@ func (a *KinesisAdapter) Stream(logstream chan *router.Message) {
 
 		if d, ok := a.findDrainer(streamName); ok {
-			logErr(d.Buffer.Add(m))
+			go d.Buffer.Add(m)
```
Spinning up a new goroutine like this for every message might be potentially dangerous:
1. We won't have control over applying any backpressure so if Add starts falling behind, this could spin up a large number of goroutines.
2. I _think_ there's no guarantee that the goroutines will be executed in the same order that they're started, so log lines might get added out of order.

I wonder if it would be better to treat individual buffers as short lived, where you add messages to the buffer until it's full, then send that buffer off on a channel to get flushed by another goroutine, and allocate a new buffer for the next messages. This is basically what we do in the events service: https://github.com/remind101/r101-events/blob/master/handlers/redshift/receiver.go#L89-L101 (kind of a bad example because the write still blocks any adds).

It's potentially more allocations, but probably not enough to worry about it.

benguillet committed:
Refactoring: stream.

benguillet committed:
WIP.

benguillet committed:
Add sublime files to gitignore.

benguillet committed:
Refactoring: streamer and buffer implementation.

benguillet committed:
Add stream already created test.

benguillet committed:
Mocking improvements. Add the Status function.

benguillet committed:
Add stream creation test.

benguillet committed:
Add mutex on fakeClient.

benguillet committed:
Better errors.

benguillet committed:
Replace readyWrite mutex by a channel.

benguillet committed:
Fix comment about RecordSizeLimit error.

benguillet committed:
Allow nil templates.

benguillet committed:
Implement stream tagging.

benguillet committed:
Remaining empty templates.

benguillet committed:
Fix templating.

benguillet committed:
Add a flusher test.

benguillet committed:
Fix empty template.

benguillet committed:
Implement proper dropBufferFunc.

benguillet committed:
Hold the kinesis.PutRecordsInput object itself, not a reference to it.

So when we reset it, and try to access it later, we actually access
a copy of it.

benguillet committed:
No need for the count() and input() functions.

Access direct the struct fields.

benguillet committed:
Rename buffer fields.

benguillet committed:
Add and cleanup debug logging.

benguillet committed:
Cleanup old files.

benguillet committed:
Fix flusher logging.

benguillet committed:
Log when stream is ready.

benguillet committed:
Move functions around.

benguillet committed:
Better logging consistent. Misc improvements.

benguillet committed:
Clarification about what logspout-kinesis do for you not.

benguillet committed:
Adding comments.

benguillet committed:
Remove package kineprod, since it’s not extractable as is.

benguillet committed:
Update Godeps.

benguillet committed:
Add circle.yml.

benguillet committed:
Add CI status badge.

bmarini commented on buffer.go:
```diff
@@ -11,142 +10,99 @@ import (
+		"PutRecordsLimit":     PutRecordsLimit,
+		"PutRecordsSizeLimit": PutRecordsSizeLimit,
+		"RecordSizeLimit":     RecordSizeLimit,
 	}
```
I'd go with regular struct fields or perhaps a sub struct instead of a map. No need for a map since the keys never change, plus a little more error prone and less memory efficient.

benguillet replied:
Yeah, duh. Fixed. Not sure what i was thinking there.

benguillet committed:
map -> struct. Duh.

ejholmes commented on buffer.go:
```diff
@@ -11,142 +10,103 @@ import (
+	pKey, err := executeTmpl(b.pKeyTmpl, m)
 	if err != nil {
-		return err
+		panic(err)
```
Why panic and not return the err here?

benguillet replied:
Errors just get logged (actually not right now on the call to `w.buffer.add()` but im just gonna update that. But if we can't get any partition key we have a big problem, so we should crash. Or report to honeybadger.. but that's for another pr.

ejholmes replied:
I talk about this in another comment, but I think we really don't want to crash, because that will result in lost log lines for other adapters that were behaving correctly. We should definitely be loud about the error, but crashing and restarting won't do much good since we'll just keep crashing most likely.

ejholmes commented on buffer.go:
```diff
@@ -11,142 +10,103 @@ import (
-func (r *recordBuffer) Add(m *router.Message) error {
-	data := m.Data
-	dataLen := len(data)
+func (b *buffer) add(m *router.Message) error {
```
Is add ever called from multiple goroutines? Does this need a mutex?

benguillet replied:
No, it's only called by the bufferMessages goroutine, and then we sent a copy of the buffer to `w.buffers`. I wonder if we could lose a message if the add() hasn't finished by the time a `reset()` is triggered by the ticker though...

ejholmes commented on stream.go:
```diff
@@ -0,0 +1,126 @@
+	select {
+	case <-s.readyWrite:
+		s.ready = true
+		log.Printf("ready! stream: %s", s.name)
```
Should this be `debug()`?

benguillet replied:
No i want to log this line everytime, even when `KINESIS_DEBUG` isn't set.

ejholmes commented on kinesis.go:
```diff
@@ -14,86 +13,90 @@ func init() {
+
+			s := NewStream(sn, tags, a.PKeyTmpl)
+			s.Start()
+			s.Writer.Start()
```
Might want to make s.Start() call s.Writer.Start().

benguillet replied:
I like having more granularity over what it's getting started, especially for the tests.

ejholmes commented on stream.go:
```diff
@@ -0,0 +1,126 @@
+
+func (s *Stream) tag() {
+	// wait to be created...
+	<-s.readyTag
```
might want to make this a `chan error` and check if err isn't nil before tagging.

ejholmes commented on stream.go:
```diff
@@ -0,0 +1,126 @@
+		Tags:       *s.tags,
+	})
+	if !tagged {
+		panic(err)
```
We probably shouldn't panic here. We could get into a state where logspout would just continually crash.

Might want to make `readWrite` a `chan error`, so we can check if there's an error. If there is, then any calls to `Write` should return that error.

ejholmes commented on writer.go:
```diff
@@ -0,0 +1,80 @@
+func (w *writer) bufferMessages() {
+	flush := func() {
+		select {
+		case w.buffers <- *w.buffer:
```
`flusher` only ever accesses `buffer.input`, so it should probably take a `*kinesis.PutRecordsInput` instead.

I think we should also move all of the logic for dropping into the `flusher` implementation, which makes writer a lot simpler since you can just do:

``` go
if w.buffer.full(m) {
        w.flusher.flush(w.buffer.input)
}
```

ejholmes commented on kinesis.go:
```diff
@@ -14,86 +13,90 @@ func init() {
+	for m := range logstream {
+		sn, err := executeTmpl(a.StreamTmpl, m)
 		if err != nil {
 			logErr(err)
```
We should make logErr a public variable inside this package. Then we can set it from [honeybadger.go](https://github.com/remind101/logspout/blob/master/honeybadger.go) to report the error to honeybadger:

``` go
var ErrorHandler = logErr
```

``` go
kinesis.ErrorHandler = func(err error) {
        honeybadger.Notify(err)
}
```

benguillet replied:
yes!

ejholmes commented on stream.go:
```diff
@@ -0,0 +1,126 @@
+)
+
+type ErrStreamNotReady struct {
+	s string
```
Super minor, but in general it's nice to make fields like this on Error structs exported, so consumers can type assert and check individual fields:

``` go
ErrStreamNotReady struct {
        Stream string
}
```

ejholmes commented on stream.go:
```diff
@@ -0,0 +1,126 @@
+	client     Client
+	name       string
+	tags       *map[string]*string
+	Writer     *writer
```
Also minor, should just make this unexported. We probably wouldn't want consumers to use this directly (same package, so consumers still can, but still).

ejholmes commented on client.go:
```diff
@@ -0,0 +1,48 @@
+	return *resp.StreamDescription.StreamStatus
+}
+
+func (c *client) Tag(input *kinesis.AddTagsToStreamInput) (bool, error) {
```
Seems unnecessary to return `(bool, error)` instead of just `error`. `err != nil` does the same thing.

ejholmes commented on stream.go:
```diff
@@ -0,0 +1,126 @@
+// Start runs the goroutines making calls to create and tag the stream on
+// AWS.
+func (s *Stream) Start() {
+	go s.create()
```
You could actually get rid of some of these channels and just make it something as simple as:

``` go
func (s *Stream) Write(m *router.Message) error {
        switch {
        case s.err != nil:
                return s.err
        case s.ready:
                s.Writer.Write(m)
                return nil
        default:
                return &ErrStreamNotReady{s: s.name}
        }
}

func (s *Stream) Start() {
        go s.start()
}

func (s *Stream) start() {
        if s.err = s.create(); s.err != nil {
                return
        }

        if s.err = s.tag(); s.err != nil {
                return
        }

        s.ready = true
}
```

It would be nice if we buffered up messages while the stream was in a pending state but probably not for this PR.

ejholmes commented:
Aside from those comments, this looks pretty good to me.

It might make sense from a performance standpoint to eventually have a separate writer per partition key/container. Since we only need to guarantee ordering of log lines for an individual container, and the containers partition key would never change, we could spin up a separate writer/flusher in a go routine for each container. This means that goroutines would grow unbounded unless we could get events from logspout when a container dies though...

benguillet committed:
Better error reporting.

benguillet committed:
Simplify stream creation and tagging.

benguillet committed:
Let the stream starts the writer.

benguillet committed:
Export error fields.

benguillet committed:
Move dropping responsibility to the flusher.

benguillet committed:
Add flusher tests.

benguillet committed:
Fix gometalinter warnings.

ejholmes commented:
:shipit:
